### PR TITLE
chore(tracing): support span links for v0.4 (default) trace encoding

### PR DESCRIPTION
**Pull request by mabdinur**
Currently span links are only serialized if ``DD_TRACE_API_VERSION`` is set to v0.5. Since the default trace api version is v0.4 this feature is NOT supported out of the box. 

This PR adds support for serializing span links in the v0.4 format.

Note - Support for span links is undocumented. We will officially announce this feature in a release note in a future change. This PR does not require a release note.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/internal/_encoding.pyx [hunk 1]
```diff
@@ -547,6 +550,50 @@ cdef class MsgpackEncoderV03(MsgpackEncoderBase):
...
+                    ret = msgpack_pack_map(&self.pk, len(attributes))
+                    for attr_k, attr_v in attributes:
+                        ret = pack_text(&self.pk, attr_k)
+                        if ret != 0:
+                            return ret
+                        ret = pack_text(&self.pk, attr_v)
+                        if ret != 0:
+                            return ret
+                else:
+                    raise ValueError(f"{v} has an unsupported type, type={type(v)}")
```
Comment (mabdinur) [inline ddtrace/internal/_encoding.pyx:592]:
```suggestion
                    raise ValueError(f"span link field ({k}) has a value ({v}) with an unsupported type, type={type(v)}")
```

File: ddtrace/internal/_encoding.pyx [hunk 2]
```diff
@@ -547,6 +550,50 @@ cdef class MsgpackEncoderV03(MsgpackEncoderBase):
...
+                    ret = msgpack_pack_map(&self.pk, len(attributes))
+                    for attr_k, attr_v in attributes:
+                        ret = pack_text(&self.pk, attr_k)
+                        if ret != 0:
+                            return ret
+                        ret = pack_text(&self.pk, attr_v)
+                        if ret != 0:
+                            return ret
+                else:
+                    raise ValueError(f"span link field ({k}) has a value ({v}) with an unsupported type, type={type(v)}")
```
Comment (mabdinur) [inline ddtrace/internal/_encoding.pyx:592]:
```suggestion
                    raise ValueError(f"Failed to encode SpanLink. {k}={v} contains an unsupported type, {type(v)}")
```

File: tests/tracer/test_encoders.py [hunk 3]
```diff
@@ -400,6 +400,62 @@ def test_span_types(encoding, span, tags):
...
+    encoder.put([span])
+    decoded_trace = decode(encoder.encode())
+    # ensure one trace was decoded
+    assert len(decoded_trace) == 1
+    # ensure trace has one span
+    assert len(decoded_trace[0]) == 1
+
+    decoded_span = decoded_trace[0][0]
+    assert b"span_links" in decoded_span
+
```
Comment (mabdinur) [inline tests/tracer/test_encoders.py:440]:
```suggestion
```

File: tests/tracer/test_encoders.py [hunk 4]
```diff
@@ -400,6 +400,62 @@ def test_span_types(encoding, span, tags):
...
+                    "moon": "ears",
+                    "link.name": "link_name",
+                    "link.kind": "link_kind",
+                    "someval": 1,
+                    "drop_me": "bye",
+                },
+            )
+        ],
+    )
+
```
Comment (mabdinur) [inline tests/tracer/test_encoders.py:424]:
```suggestion
```

**Review by ZStriker19 — APPROVED** [review]

File: tests/tracer/test_encoders.py [hunk 5]
```diff
@@ -400,6 +400,60 @@ def test_span_types(encoding, span, tags):
...
 
 
+def test_span_link_v04_encoding():
+    encoder = MSGPACK_ENCODERS["v0.4"](1 << 20, 1 << 20)
+
+    span = Span(
+        "s1",
+        links=[
+            SpanLink(
+                trace_id=(123 << 64) + 456,
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply tests/tracer/test_encoders.py:413]: Yup that's what this line does.

(123 << 64) + 456  is equal to 2268949521066274849224 
`2268949521066274849224` is represented in binary with 71 bits. This makes it "128 bit integer".

bin(2268949521066274849224) -> `11110110000000000000000000000000000000000000000000000000000000111001000`

**Pull request merged**